### ISSUE TITLE
how should `assert_type` handle the float/complex special case?

### ISSUE THREAD

**Issue by carljm** [astral-sh/ty#197]
### Description

We follow the typing spec and other type checkers in implementing the special case that a `float` annotation means `int | float`, and a `complex` annotation means `int | float | complex`.

We don't currently follow other type checkers in trying to "hide" this special case by revealing just `float` or `complex` when the actual type we have is `int | float` or `int | float | complex`. Doing this leads to very strange-looking behaviors, like this from pyright:

```py
def _(x: float):
    reveal_type(x)  # float
    if isinstance(x, float):
        reveal_type(x)  # float
    reveal_type(x)  # float | int ??
```

But this leads us to fail this type assertion:

```py
assert_type(3.14, float)
```

Because we interpret `float` there as a type expression, meaning we take it to be `int | float`, but we actually infer just `float` for the literal.

**Comment by InSyncWithFoo** [comment]
I think `float` should be interpreted as `int | float` no matter the context. That helps with consistency.

Instead, `knot_extensions` should export `BareFloat` and `BareComplex` as public APIs for such use cases (credits go to @sharkdp):

```python
type BareFloat = TypeOf[1.0]
type BareComplex = TypeOf[1j]
```

**Comment by sharkdp** [comment]
I think we still haven't decided if `ty_extensions` should be user-facing, but for now, `ty_extensions.JustFloat` and `ty_extensions.JustComplex` are available:

```py
from ty_extensions import JustFloat

def only_float_allowed(f: JustFloat): ...

only_float_allowed(1.0)
only_float_allowed(1)  # invalid_argument-type
```
https://play.ty.dev/e83982c2-6ab6-409a-8608-97f7f0baa14a
